### PR TITLE
Closes #910: Add regex support to `Strings.flatten`:

### LINKED ISSUE #910: Add regex support to `Strings.flatten`
In support of #894 :
Add support for flattening a Strings using a regex delimiter 

Example:
```python
answer_flat = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten', '', 'eleven']
answer_map = [0, 2, 5, 6, 11]

orig = ak.array(['one|two', 'three|four|five', 'six', 'seven|eight|nine|ten|', 'eleven'])
digit = ak.array(['one1two', 'three2four3five', 'six', 'seven4eight5nine6ten7', 'eleven'])
under = ak.array(['one_two', 'three_four__five', 'six', 'seven_____eight__nine____ten_', 'eleven'])

orig_flat, orig_map = orig.flatten('|', return_segments=True)
digit_flat, digit_map = digit.flatten('\\d', return_segments=True, regex=True)
under_flat, under_map = under.flatten('_+', return_segments=True, regex=True)

# all the flats are the same and all the maps are the same
```

### PR DESCRIPTION
**Pull request by pierce314159**
Closes #910:
- Adds regex support for `Strings.flatten` and a test for the regex functionality

Example:
```python
# regex flatten functionality
>>> under = ak.array(['one_two', 'three_____four____five', 'six'])
>>> under_flat, under_map = under.flatten('_+', return_segments=True, regex=True)
>>> under_flat
array(['one', 'two', 'three', 'four', 'five', 'six'])
>>> under_map
array([0, 2, 5])

# compared to existing flatten
>>> orig = ak.array(['one|two', 'three|four|five', 'six'])
>>> flat, map = orig.flatten('|', return_segments=True)
>>> flat
array(['one', 'two', 'three', 'four', 'five', 'six'])
>>> map
array([0, 2, 5])
```

### CODE REVIEW

**Review by hokiegeek2 — APPROVED** [review]

**Comment by reuster986** [comment]
I like this!

I'm wondering if it makes sense to add a benchmark for this function, like for the substring searches. I'm open to it being a separate PR, whatever you think best.

**Comment by pierce314159** [comment]
@reuster986 Sorry for the delay, I added a benchmark comparing non-regex flatten, regex flatten with a literal delimiter, and regex flatten with a pattern delimiter. Let me know if this looks like what you had in mind

**Review by reuster986 — APPROVED** [review]
Looks good to me! We can look at benchmark output after we merge.

File: benchmarks/graph_infra/arkouda.graph [hunk 1]
```diff
@@ -94,6 +94,12 @@ files: substring_search.dat, substring_search.dat, substring_search.dat
 graphtitle: Substring Search Performance
 ylabel: Performance (GiB/s)
 
+perfkeys: non-regex flatten with literal delimiter Average rate =, regex flatten with literal delimiter Average rate =, regex flatten with pattern delimiter Average rate =
+graphkeys: non-regex flatten with literal delimiter GiB/s, regex flatten with literal delimiter GiB/s, regex flatten with pattern delimiter GiB/s
+files: flatten.dat, flatten.dat, flatten.dat
+graphtitle: Flatten Performance
+ylabel: Performance (GiB/s)
+
```
Comment (ronawho) [inline benchmarks/graph_infra/arkouda.graph:102]: This is minor and only impacts what suite the generated graphs end up in, but could you move the substring/flatten graphs into `arkouda-string.graph`?

**Pull request merged**